source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -0,0 +1,433 @@
+
+void HappyEyeballsConnectionImpl::connect() {
+  ASSERT(!connect_finished_);
+  connections_[0]->connect();
+  maybeScheduleNextAttempt();
+}
+
+void HappyEyeballsConnectionImpl::addWriteFilter(WriteFilterSharedPtr filter) {
+  ASSERT(connect_finished_);
+  connections_[0]->addWriteFilter(filter);
+}
+
+void HappyEyeballsConnectionImpl::addFilter(FilterSharedPtr filter) {
+  ASSERT(connect_finished_);
+  connections_[0]->addFilter(filter);
+}
+
+void HappyEyeballsConnectionImpl::addReadFilter(ReadFilterSharedPtr filter) {
+  if (connect_finished_) {
+    connections_[0]->addReadFilter(filter);
+    return;
+  }
+  post_connect_state_.read_filters_.push_back(filter);
+}
+
+void HappyEyeballsConnectionImpl::removeReadFilter(ReadFilterSharedPtr filter) {
+  ASSERT(connect_finished_);
+  connections_[0]->removeReadFilter(filter);
+}
+
+bool HappyEyeballsConnectionImpl::initializeReadFilters() {
+  ASSERT(connect_finished_);
+  return connections_[0]->initializeReadFilters();
+}
+
+void HappyEyeballsConnectionImpl::addBytesSentCallback(Connection::BytesSentCb cb) {
+  ASSERT(connect_finished_);
+  connections_[0]->addBytesSentCallback(cb);
+}
+
+void HappyEyeballsConnectionImpl::enableHalfClose(bool enabled) {
+  ASSERT(connect_finished_);
+  connections_[0]->enableHalfClose(enabled);
+}
+
+bool HappyEyeballsConnectionImpl::isHalfCloseEnabled() {
+  ASSERT(connect_finished_);
+  return connections_[0]->isHalfCloseEnabled();
+}
+
+std::string HappyEyeballsConnectionImpl::nextProtocol() const {
+  ASSERT(connect_finished_);
+  return connections_[0]->nextProtocol();
+}
+
+void HappyEyeballsConnectionImpl::noDelay(bool enable) {
+  if (!connect_finished_) {
+    per_connection_state_.no_delay_ = enable;
+  }
+  for (auto& connection : connections_) {
+    connection->noDelay(enable);
+  }
+}
+
+void HappyEyeballsConnectionImpl::readDisable(bool disable) {
+  ASSERT(connect_finished_);
+  connections_[0]->readDisable(disable);
+}
+
+void HappyEyeballsConnectionImpl::detectEarlyCloseWhenReadDisabled(bool value) {
+  if (!connect_finished_) {
+    per_connection_state_.detect_early_close_when_read_disabled_ = value;
+  }
+  for (auto& connection : connections_) {
+    connection->detectEarlyCloseWhenReadDisabled(value);
+  }
+}
+
+bool HappyEyeballsConnectionImpl::readEnabled() const {
+  if (!connect_finished_) {
+    ASSERT(connections_[0]->readEnabled());
+  }
+  return connections_[0]->readEnabled();
+}
+
+const SocketAddressProvider& HappyEyeballsConnectionImpl::addressProvider() const {
+  ASSERT(connect_finished_);
+  return connections_[0]->addressProvider();
+}
+
+SocketAddressProviderSharedPtr HappyEyeballsConnectionImpl::addressProviderSharedPtr() const {
+  ASSERT(connect_finished_);
+  return connections_[0]->addressProviderSharedPtr();
+}
+
+absl::optional<Connection::UnixDomainSocketPeerCredentials>
+HappyEyeballsConnectionImpl::unixSocketPeerCredentials() const {
+  ASSERT(connect_finished_);
```

Comment:
optional: I wonder if all the connect_finished assers should be ENVOY_BUG so we have a better chance of catching in prod.

Reply:
Ooo! I didn't realize there was such a thing. Done!

Reply:
This is now moot.